expression/constant.go

```diff
@@ -99,11 +99,10 @@ func (c *Constant) Eval(_ chunk.Row) (types.Datum, error) {
 				c.Value.SetNull()
 				return c.Value, nil
 			}
-			retType := types.NewFieldType(c.RetType.Tp)
-			if retType.Tp == mysql.TypeUnspecified {
-				retType.Tp = mysql.TypeVarString
+			if c.RetType.Tp == mysql.TypeUnspecified {
```

Comment:
When will `c.RetType.Tp == mysql.TypeUnspecified` be true?

Reply:
Here are the example for the case, which is the snippet in ``TestPreparedIssue7579()`` of ``tidb/executor/prepared_test.go``:     

```
create table t (a int, b int, index a_idx(a))
prepare stmt from 'select a, b from t where ? order by b asc'
set @param = true
execute stmt using @param;
```
Here, the value of the parameter ``?`` is ``mysql.TypeUnspecified``.

Reply:
@dbjoa Maybe it's better to do this after the `c.DeferredExpr` is constructed? For example, https://github.com/pingcap/tidb/pull/8619/files#diff-33ebff1614e7db77b69dd02c2eee26c2R1245

Reply:
@zz-jason Thank you for the comment. However, I think that ``types.DefaultParamTypeForValue()`` might be the right place to change the parameter's return type to be ``mysql.TypeVarString`` from ``mysql.TypeUnspecified`` like the updated PR: 

https://github.com/pingcap/tidb/blob/5617f84d10ab678594041f59f0a0a89b024c9096/types/field_type.go#L129-L141

Reply:
Oh, yeah, this is the ideal place 😄